GVFS/GVFS.Common/NuGetUpgrade/NuGetUpgrader.cs

```diff
@@ -492,6 +522,71 @@ private bool TryRecursivelyDeleteInstallerDirectory(out string error)
             return true;
         }
 
+        private IList<IPackageSearchMetadata> QueryFeedWithRetry()
+        {
+            int retryCount = 0;
+            IList<IPackageSearchMetadata> queryResults = null;
+
+            do
+            {
+                try
+                {
+                    queryResults = this.nuGetFeed.QueryFeedAsync(this.nuGetUpgraderConfig.PackageFeedName).GetAwaiter().GetResult();
+                    return queryResults;
+                }
+                catch (Exception ex) when (ex is FatalProtocolException &&
+                                           ex.InnerException is System.Net.Http.HttpRequestException &&
+                                           (ex.InnerException.Message.Contains("401") || ex.InnerException.Message.Contains("403")) &&
+                                           retryCount == 0)
+                {
+                    // If we fail to query the feed due to an authorization error, then it is possible we have stale
+                    // credentials, or credentials without the correct scope. Re-aquire fresh credentials and try again.
+                    EventMetadata data = CreateEventMetadata(ex);
+                    this.tracer.RelatedWarning(data, "Failed to query feed due to unauthorized error. Re-acquiring new credentials and trying again.");
+
+                    retryCount++;
+
+                    if (!this.TryRefreshCredentials(out string error))
+                    {
+                        // If we were unable to re-acquire credentials, throw a new exception indicating that we tried to handle this, but were unable to.
+                        throw new Exception($"Failed to query the feed for upgrade packages due to: {ex.Message}, and was not able to re-acquire new credentials due to: {error}", ex);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    EventMetadata data = CreateEventMetadata(ex);
+                    string message = $"Error encountered when querying NuGet feed. Number of attempts: {retryCount + 1}";
+                    this.tracer.RelatedWarning(data, message);
+                    throw new Exception($"Failed to query the NuGet package feed due to error: {ex.Message}", ex);
+                }
+            }
+            while (true);

```

Comment:
Are we retrying forever if something doesn't go as expected?  I would prefer some sort of retry limit.

Reply:
We will only retry 1 time - I can restructure to make this more explicit / obvious

Reply:
Yeah I would make a method and check a return code and call the method again if it failed rather than a loop.